Implement flexible shape/dims/size API

michaelosthege opened the pull request:
## Terminology
+ _implied_ dimensions are those that a RV takes without specifying any of `shape,dims,size,observed`.
+ The `size` of an RV are all dimensions _in addition to_ (preceeding) the _implied_ dimensions.

## Key Takeaways
+ `shape` can be numeric/symbolic and supports `Ellipsis` in the last position. Without `Ellipsis` a `SpecifyShape` `Op` is added automatically.
+ `dims` can be `None` and are defined at first mention. `Ellipsis` is also supported. In combination with `pm.Data` the dimensions specified through `dims` are re-sizeable.
+ `size` does care about implied dimensions in any way and just adds new dimensions to the RV. Internally `size` is inferred from `shape` and `dims`.
+ No more resizing in `Model.make_obs_var` because it leaves code in `Distribution.__new__` not knowing the output shape.

## ToDo
+ [x] `Distribution.dist()` and `Distribution.__new__` are similar in use (`dims` and `observed` only through `__new__`)
+ [x] Inputs are checked and informative errors are raised where possible
+ [x] Release notes were updated
+ [x] Some tests were edited to use `size`/`shape` appropriately -- some tests illegally used `size` to specify size _and implied dimensions_.

## ToDo (optional)
+ [ ] Ideally, it should be possible to pass `dims` in combination with `size`. We should just require a single source of truth (dim names must be undefined for dimensions that `size` integers apply to).

OriolAbril commented on pymc3/distributions/distribution.py:
```diff
@@ -141,26 +159,102 @@ def __new__(cls, name, *args, **kwargs):
+                        f"This is more than the {implied_ndim} dimensions implied by the distribution parameters. "
+                        f"To replicate the RV beyond its implied dimensionality use `size={size}` instead.",
+                        UserWarning,
+                    )
```
Does this mean that doing `pm.Normal("a", 0, 1, shape=(3, 20))` will raise a user warning? If so, can we deprecate and/or remove `shape`? 

I have been lost for a while, I think discussing this on the meeting confused me more than it helped. I thought we would only keep one of `size`/`shape`, seeing this I think shape should be removed, but I have been partial to `dims`+`size` since the beginning.

michaelosthege replied:
Yes, your example will raise this warning.
And it's correct that with the current state of this PR the full shape specification does not add anything. My first priority was to get things working, now the next step is, for example, to add the automatic shape assert whenever `shape` is passed.

twiecki replied:
That's not what we discussed. Most of us were in agreement that explicit shapes are better than implicit size.

brandonwillard replied:
I do not recall _most_ people agreeing on anything involving this entire matter.

The point in favor of the `shape` parameter has only ever been that it's somehow more "explicit" and/or better than a comment stating the expected shape, and this point somehow outweighs the effort/know-how required to specify it correctly and the unnecessary limitations it imposes.

Remember, we could accomplish the same thing without all this extra project-level code if we simply wrote a small blurb in the documentation describing how people can use Aesara's `SpecifyShape` instead.

For example, if one wants to be explicit about a random variable's shape:
```python
import numpy as np

import aesara
import aesara.tensor as at

from aesara.tensor.shape import specify_shape

import pymc3 as pm


mu = at.vector("mu")
Sigma = at.matrix("Sigma")

X = pm.MvNormal.dist(mu, Sigma, size=(2, 3))

# This is it; one extra line that accomplishes the exact same thing as the old
# `shape` parameter
X = specify_shape(X, (2, 3, 4))
```

What does this do for us?  Not much, but here's an example:
```python
>>> X.eval({mu: np.zeros(4), Sigma: np.eye(4)})
array([[[ 0.84616091, -1.38344376, -1.22742971, -0.66795514],
        [ 0.90845087,  0.366118  ,  0.29222136, -0.75515633],
        [ 0.15882859, -0.97699321, -1.09552517, -1.24519882]],

       [[-1.05805731, -2.63053756, -0.30468915,  1.21047848],
        [-1.10110766, -0.93431473, -0.00385619, -0.97005804],
        [ 1.07159636, -1.02907491, -0.39692291,  0.39597758]]])

>>> X.eval({mu: np.zeros(2), Sigma: np.eye(2)})
...
AssertionError: SpecifyShape: dim 2 of input has shape 2, expected 4.
...
```

Unfortunately, this triviality is taking precious developer time and preventing us from actually moving forward (e.g. merging into `master`), so we need consider whether or not this `shape` thing is really more important than any or all of the implemented, in-progress, and planned `v4` improvements.

OriolAbril commented on pymc3/distributions/distribution.py:
```diff
@@ -141,26 +159,102 @@ def __new__(cls, name, *args, **kwargs):
+                raise NotImplementedError("Ellipsis-based dims inference is not implemented.")
+                dims = tuple(dims[:-1]) + implied_dims
+            dimshape = tuple(len(model.coords[dname]) for dname in dims)
+            size = tuple(dimshape[: len(dimshape) - implied_ndim])
```
can we cut the tuple on `dims` already so only the dims that are eventually passed to `size` are accessed from coords? We don't really need nor use those implied dims so we can _allow_ users to define them at sample time. Everything should still work the same if they are defined and passed already when defining the model, it just wouldn't be necessary for them to be there.

michaelosthege replied:
In principle yes, but I didn't know what happens if the `dims` passed to `model.register_rv` are incomplete.

brandonwillard commented on pymc3/distributions/distribution.py:
```diff
@@ -141,26 +159,102 @@ def __new__(cls, name, *args, **kwargs):
+            dimshape = tuple(len(model.coords[dname]) for dname in dims)
+            size = tuple(dimshape[: len(dimshape) - implied_ndim])
+        else:
+            raise Exception("This should have been unreachable code.")
```
If this is truly unreachable, then it isn't necessary; otherwise, there's something missing in this logic that needs to be addressed directly

brandonwillard commented on pymc3/distributions/distribution.py:
```diff
@@ -141,26 +159,102 @@ def __new__(cls, name, *args, **kwargs):
+            size = ()
+        elif size is not None:
+            # User already specified how to expand the RV
+            pass
```
There shouldn't be a condition for a no-op.

michaelosthege replied:
> This introduces a considerable amount of complexity into `Distribution.__new__`; the `shape` logic really needs its own function/method, at the very least.

Absolutely. This is still a draft and the `.dist()` API is also on the ToDo list. For that reason too, the logic will move into a helper function.

> Also, this PR unnecessarily reinstates the use of `shape` internally (e.g. numerous change from `size` to `shape` in the tests), which is the **opposite** of what we set out to do.

Many tests trigger `UserWarning`s or use the API in another discouraged fashion. Where it makes sense, they should be refactored to use the API in a best-practice way.
In many tests that will still be `shape`, because they also pass a `testval`, making the test shape-explicit.
I didn't want to invest that time until we have pinned down what exactly are the unsupported/discouraged API flavors.

brandonwillard commented on pymc3/distributions/distribution.py:
```diff
@@ -141,26 +159,102 @@ def __new__(cls, name, *args, **kwargs):
+        if size is not None and Ellipsis in size:
+            raise ValueError("The `size` parameter cannot contain an Ellipsis. Actual: {size}")
+        if not _valid_ellipsis_position(shape):
+            raise ValueError(
```
This is an excessive amount of parameter validation.  If it's absolutely necessary, it should be delegated to a validation function/method.

brandonwillard commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -174,12 +174,7 @@ def get_random_variable(self, shape, with_vector_params=False, name=None):
-                        else:
-                            size = shape[:-ndim_supp]
-                        return self.distribution(name, size=size, transform=None, **params)
+                        return self.distribution(name, shape=shape, transform=None, **params)
```
Remember, we are **not** using `shape` internally.

michaelosthege replied:
Yes. I reverted these to `shape` because I was lazy. They should be refactored to use `shape`, but for many test that means we have to `pytest.mark.parametrize` with both `shape` and `size`, because we often use the `shape` for asserts.

brandonwillard reviewed (changes requested):
This introduces a considerable amount of complexity into `Distribution.__new__`; the `shape` logic really needs its own function/method, at the very least.

Also, this PR unnecessarily reinstates the use of `shape` internally (e.g. numerous change from `size` to `shape` in the tests), which is the **opposite** of what we set out to do.

michaelosthege commented:
Caution: The API from `aesara.tensor.random.basic` `Op`s such as `normal` resembles the NumPy API:
```python
x = aesara.shared(np.array([10,20,30,40]))
n = atr.normal(loc=x, scale=0.1, size=(2,4))     # size=(2,) does not work!
```

In PyMC3 this will equivalent to "inflexible" use of `shape`:
```python
with pm.Model():
    x = aesara.shared(np.array([10,20,30,40]))
    pm.Normal("n1", mu=x, sd=0.1, shape=(2,4))
    pm.Normal("n2", mu=x, sd=0.1, size=(2,))
```

michaelosthege commented:
With the latest push the `Distribution()` and `Distribution.dist()` APIs are now aligned and validation/warning/casting logic moved to a helper function.

`dims` are not available through the `.dist()` API until dimension name broadcasting is implemented at the Aesara `Variable` level.

Passing a non-Ellipsis shape like `shape=(1, 2, 3, 4)` is the **backwards-compatible** use of the API that, at the same time results in the automatic addition of `specify_shape`.
The `SpecifyShape` `Op` couldn't deal with scalar shapes `()`, so the corresponding tests will fail until the fixes from https://github.com/pymc-devs/aesara/pull/367 are available.

A `BestPracticeWarning` is emitted in the following scenarios:
- `pm.Normal(..., shape=3)` → "The `shape` parameter should be a list or tuple."
- `pm.Normal(..., dims="country")` → "The `dims` parameter should be a list or tuple."

No `BestPracticeWarnings` are emitted in the following scenarios:
- `pm.Normal(..., mu=[1,2], shape=(3, 2))` → Automatic `specify_shape` will alert the user when the implied shape changes.
- `pm.Normal(..., mu=[1,2], shape=(3, ...))` → Notation indicates that there are more than 1 dims (debatable if we should nudge towards `size`).


Before we invest too much time into `shape`/`size` arguments, I would like to point out that `dims` are 10x more useful, because they make working with the resulting `InferenceData` much easier.

OriolAbril commented:
> Before we invest too much time into shape/size arguments, I would like to point out that dims are 10x more useful, because they make working with the resulting InferenceData much easier.

:100:

OriolAbril commented on pymc3/distributions/distribution.py:
```diff
@@ -141,40 +248,114 @@ def __new__(cls, name, *args, **kwargs):
-        if "shape" in kwargs:
-            raise DeprecationWarning("The `shape` keyword is deprecated; use `size`.")
+            dimshape = tuple(len(model.coords[dname]) for dname in dims)
+            size = tuple(dimshape[: len(dimshape) - rv_out.ndim])
```
```suggestion
            size = tuple(len(model.coords[dname]) for dname in dims[: len(dimshape) - rv_out.ndim])
```

I think we should use this and not require all coords to be present in the model context. Conceptually we don't need them to be defined until conversion to ArviZ time, and now with size practice matches theory :tada: and they are indeed not used anymore, we get their length only to discard that info in the next line.

michaelosthege replied:
Sounds good, but since we know the implied `ndim` we can also substitute the missing dim names with `f"{name}_dim_{d}"`. That should at least give us better chances of not getting downstream issues, right?

OriolAbril replied:
Not sure I follow what you mean by missing dim names, we still need to pass the dim names for the implied dims right? I don't think the `: len(dimshape) - rv_out.ndim` would work otherwise. 

If we want to be extra careful, we can check that the number of dimension names provided matches the number of dimensions of the variable. I'm not sure how we'd go about that because we can use dims to extend the implied dimensions, but in any case, we don't need their lengths for that.

If the concern is about the number of implied dimensions changing afterwards for a 2nd round of sampling, posterior predictive sampling... as long as the number of dimensions increases, ArviZ will handle that automatically and generate the `rv_name_dim_#` alone. If they are reduced it will error out, but I'd say that's on the user doing that, if they modify the number of dims they should then pass the new `dims` to ArviZ via `idata_kwargs`, `dims` or whatever. Example:

```python
import numpy as np
import arviz as az

data = np.random.normal(size=(4, 500, 5, 3, 7))
idata = az.from_dict({"a": data}, dims={"a": ["subject"]})
idata.posterior
# Dimensions:  (a_dim_1: 3, a_dim_2: 7, chain: 4, draw: 500, subject: 5)
# Coordinates:
#   * chain    (chain) int64 0 1 2 3
#   * draw     (draw) int64 0 1 2 3 4 5 6 7 8 ... 492 493 494 495 496 497 498 499
#   * subject  (subject) int64 0 1 2 3 4
#   * a_dim_1  (a_dim_1) int64 0 1 2
#   * a_dim_2  (a_dim_2) int64 0 1 2 3 4 5 6
# Data variables:
#     a        (chain, draw, subject, a_dim_1, a_dim_2) float64 0.3243 ... -0.8625
```

if you want to be cryptic and rely on default dim names you can even do:

```python
idata = az.from_dict({"a": data}, dims={"a": ["subject", "dose"]}, coords={"a_dim_2": [l for l in "abcdefg"]})
idata.posterior
# Dimensions:  (dose: 3, a_dim_2: 7, chain: 4, draw: 500, subject: 5)
# Coordinates:
#   * chain    (chain) int64 0 1 2 3
#   * draw     (draw) int64 0 1 2 3 4 5 6 7 8 ... 492 493 494 495 496 497 498 499
#   * subject  (subject) int64 0 1 2 3 4
#   * dose     (dose) int64 0 1 2
#   * a_dim_2  (a_dim_2) <U1 'a' 'b' 'c' 'd' 'e' 'f' 'g'
# Data variables:
#     a        (chain, draw, subject, dose, a_dim_2) float64 0.3243 ... -0.8625
```

brandonwillard commented:
> `dims` are not available through the `.dist()` API until dimension name broadcasting is implemented at the Aesara `Variable` level.

To be clear, this is not a change that we should make in Aesara.  Dimension names are completely unrelated to the concerns of Aesara and we absolutely must keep the objects in that library as simple and efficient as possible.

That does not mean that we cannot consider creating a custom `TensorVariable` or `TensorType` class for this purpose in PyMC3; however, I would still hesitate to do that unless it's absolutely necessary.

For instance, why can't we just use `Variable.tag` to hold these dimension labels and reason with/about them in PyMC3?

brandonwillard commented:
> A `BestPracticeWarning` is emitted in the following scenarios:
> 
> * `pm.Normal(..., shape=3)` → "The `shape` parameter should be a list or tuple."
> * `pm.Normal(..., dims="country")` → "The `dims` parameter should be a list or tuple."

These are superfluous and oddly strict.  NumPy doesn't even complain about this; why should PyMC?

twiecki commented:
> > A `BestPracticeWarning` is emitted in the following scenarios:
> > 
> > * `pm.Normal(..., shape=3)` → "The `shape` parameter should be a list or tuple."
> > * `pm.Normal(..., dims="country")` → "The `dims` parameter should be a list or tuple."
> 
> These are superfluous and oddly strict. NumPy doesn't even complain about this; why should PyMC?

I agree, I don't think a warning is required here.

michaelosthege commented:
> I agree, I don't think a warning is required here.

Okay, this leaves us with no `BestPracticeWarning` at all.

`dims=("town", ...)` now works too. The `...`-implied dims are set to he default pattern of `f"{name}_dim_{d}"` where `d` counting includes the `size` dimensions, like it would if there weren't any dim names at all.
For those implied dim names, `None` is set in the `Model.RV_dims[rv_name]` tuple of coordinate values. @OriolAbril is that something we can work with downstream, or would you prefer a different appraoch?

OriolAbril commented on pymc3/distributions/distribution.py:
```diff
@@ -141,40 +245,116 @@ def __new__(cls, name, *args, **kwargs):
+        # without a modelcontext and dims are not yet tracked at the Aesara level.
+        if dims is not None:
+            # Infer size from dims (and coords)
+            size = tuple(len(model.coords[dname]) for dname in dims if dname != Ellipsis)
```
I think this won't achieve the goal of not requiring implied dims to have their coord values passed to `pm.Model(coords=coords)`. Take a linear regression for example:

```python
with pm.Model() as penguins:
    x = pm.Data("x", df[["bill_length_mm", "bill_depth_mm", "flipper_length_mm"]], dims=("covariate", "obs_id"))
    obs_mass = pm.Data("obs_mass", df["mass"], dims="obs_id")

    beta = pm.Normal("beta", mu=at.zeros(x.shape[0]), dims="covariate")
    sigma = pm.HalfNormal("sigma", 2)
    mass = pm.Normal("mass", mu = x @ beta, sigma=sigma, dims="obs_id", observed=obs_mass)
```

If I understood everything correctly with the improved shape handling in `v4`, having defined a code like the above I'll be able to do any of the following:

```python
with penguins:
    idata = pm.sample(idata_kwargs={"coords": {
        "covariate": ["bill_length_mm", "bill_depth_mm", "flipper_length_mm"]
    }})
    # would also work without passing any coords at all, 
    # but then `covariate` dim would be indexed by integers

# add a new covariate and an intercept to the model
covariate_names = ["ones", "bill_length_mm", "bill_depth_mm", "flipper_length_mm", "sex"]
with penguins:
    pm.set_data({"x": df[covariate_names]})
    idata = pm.sample(idata_kwargs={"coords": {"covariate": covariate_names}})

# cross validation
for idx in cv_idx_or_slices:
    df_i = df.loc[idx, :]
    # here probably doing something like 
    # penguins.set_coords({"covariate": covariate_names}) 
    # would be more convenient, but still no `obs_id` in the model
    with penguins:
        pm.set_data({"x": df_i[covariate_names], "obs_mass": df_i["mass"]})
        idata_i = pm.sample(idata_kwargs={"coords": {"covariate": covariate_names}})
```

I know I can "overwrite" the coordinates in `model.coords` using `idata_kwargs`, but:
* why should I provide coords at all if like with `obs_id` it is an implied dim and I'm already fine with coord values being integers?
* why should I provide coords _at model creation_ if I plan to modify those coords every time I fit the model? (i.e. to explore different combinations of covariates)

OriolAbril replied:
I think this line can be moved inside the `if` in order to allow this usecase, I personally will mostly use dims without ellipsis (I already use `dims` in deterministics only for explicitness sake and to get everything converted nicely to ArviZ without extrra args) and 1) would love doing that 2) I think it's powerful and conceptually clear feature.

michaelosthege replied:
Not sure I got your point completely.
> why should I provide coords at all if like with obs_id it is an implied dim and I'm already fine with coord values being integers?
> why should I provide coords at model creation if I plan to modify those coords every time I fit the model? (i.e. to explore different combinations of covariates)

You shouldn't need to do any of the above. But in `Distribution.__new__` the length of the implied dimensions is not necassarily fixed (right?), so that's not the place where to auto-define the integer coords.
That's why my question is more like: "_Can we auto-create coords missing for individual dimensions (where the entry `is None`) at the time of `InferenceData` conversion?_"

With `pm.Data` I would propose to add a `coords` dict kwarg to `set_value` so swapping data and the corresponding coords is the same command.

OriolAbril replied:
> You shouldn't need to do any of the above

The issue is that I currently have to, because I am using dim names for implied dimensions and the current code will try to do `len(model.coords[dname])` so I need to provide those coords in `pm.Model`. 

> But in Distribution.__new__ the length of the implied dimensions is not necassarily fixed (right?)

I think so

> so that's not the place where to auto-define the integer coords.
That's why my question is more like: "Can we auto-create coords missing for individual dimensions (where the entry is None) at the time of InferenceData conversion?"

This is more related to my comment below, we shouldn't auto-create coords at all, ever. ArviZ will automatically handle that. In fact, we shouldn't even need to auto create dim names either, both these tasks are basically duplicating ArviZ code in a place where it should not be so we need to resort to tricks like using `None`. As I commented below, the way to indicate that some coordinate values should be auto-created is by the dim name not being present in the coords dict.

> With pm.Data I would propose to add a coords dict kwarg to set_value so swapping data and the corresponding coords is the same command.

This sounds great

michaelosthege replied:
> [...] and the current code will try to do len(model.coords[dname]) so I need to provide those coords in pm.Model

Oh now I get it!
Okay, here's the challenge:

We don't know how many dimensions were _not_ passed unless `Ellipsis` is used.
Without `Ellipsis` we must assume that `dims` corresponds to all dimensions.

To figure out `size`, at least the first `len(dims) - ndim_implied` `dims` _must_ be known to the model.
Unless `size` can be symbolic (not sure about that), in which case we must find a `SharedVariable` corresponding to the dimension name, from which we can symbolically get the length of size-dims.

brandonwillard replied:
Yes, `size` can definitely be symbolic.

brandonwillard replied:
> We don't know how many dimensions were _not_ passed unless `Ellipsis` is used.
> Without `Ellipsis` we must assume that `dims` corresponds to all dimensions.

Nothing should be made to depend on this`Ellipsis` feature.  If that's the case, then there's a serious design issue somewhere.

michaelosthege replied:
This is just saying that if there are no Ellipses in the tuple, there is not currently a good way to tell the user if they mixed up the implied dims.

With shape, a mixup would raise a somewhat informative AssertionError in the first iteration. Mixed-up dims will just crash the conversion to InferenceData.

It's really not a problem from our perspective.

OriolAbril commented on pymc3/distributions/distribution.py:
```diff
@@ -141,40 +245,116 @@ def __new__(cls, name, *args, **kwargs):
+                dims = tuple(dims[:S]) + tuple(f"{name}_dim_{d}" for d in range(S, S + rv_out.ndim))
+                for dname in dims:
+                    if not dname in model.coords:
+                        model.coords[dname] = None
```
We shouldn't care about that. If the dimension name is not in `coords` that dimension will have default coordinate values (ArviZ) or will have no coordinate values (pure xarray). Both do different things but use the same "convention" in that default/nonexistent coord values are indicated by the dim name not being a key in the coords dict. 

It's not an issue at all if not all dims have coord values defined in `coords` whereas using `None` will indicate that there are coord values (the key is present in the coords dict) and both ArviZ and xarray will try to use `None` as coordinate values and error out.

michaelosthege replied:
So the `dims` can be `("country", "town", None, None)`, but for the `None` elements the `model.coords` should remain unchanged?

Asking because we probably want a tuple of the correct length in the `model.RV_dims` dictionary: https://github.com/pymc-devs/pymc3/pull/4625/files/4c60766db3ace7765c1f5b0696cc04b70b608b25#diff-6a4ca81dc9f970173ecf8842d1f63fa74af138f33a5d88bbc6620e2e2e493a01R284

OriolAbril replied:
For ArviZ to handle that automatically, it should be `("country", "town")`. If we want to have `dims` with the same length as the variable we can remove the `None`s in `to_inferencedata`. 

I think we should go with the first option though, with no `None`s, precisely because we will use `...` to indicate that the number of dimensions can change right? Then there is no "same number of dimensions as the actual rv"

michaelosthege replied:
To the best of my knowledge the _number of dimensions_ cannot change.

Their length should be flexible:
```python
x = aesara.shared(np.ones(5))
x.set_value(np.ones(10))
```

IIRC I tried to include such a re-sizing of a shared variable in the test case for shape asserts, but it didn't work yet.

brandonwillard replied:
That's correct; the number of dimensions cannot change.

brandonwillard commented on RELEASE-NOTES.md:
```diff
@@ -8,6 +8,7 @@
 
 ### New Features
 - The `CAR` distribution has been added to allow for use of conditional autoregressions which often are used in spatial and network models.
+- The shape handling of random variables was improved dramatically (see [#4625](https://github.com/pymc-devs/pymc3/pull/4625)).
```
This isn't really a description of the changes; it's more like a statement about the quality of the (undescribed) changes.

michaelosthege replied:
I should have used "changed" instead of "improved", yes. But as you can see above our entire v4 section in the release notes is practically nonexistent and will have to be worked anyway (out of scope).

For most changes we'll probably want to link a dedicated documentation page about v3/v4 differences, that's why I didn't bother to go into detail here.

michaelosthege committed:
Allow unnamed (None) dims and undefined (None) coord values

Also refactor into properties to add docstrings and type annotations.
And no longer allow InferenceData conversion without a Model on stack.

Co-authored-by: Oriol Abril Pla <oriol.abril.pla@gmail.com>

OriolAbril commented on RELEASE-NOTES.md:
```diff
@@ -8,6 +8,11 @@
+  - With `shape` the length of dimensions must be given numerically or as scalar Aesara `Variables`.
+  - `dims` is arguably the most elegant parametrization, because it allows you to resize `pm.Data` variables and leads to well defined coordinates in `InferenceData` objects.
+  - The `size` kwarg creates new dimensions in addition to what is implied by RV parameters.
+  - An `Ellipsis` (`...`) in the last position of `shape` or `size` can be used as short-hand notation for implied dimensions. `size=(3,)` gives the same result as `shape=(3, ...)` but doesn't add a `SpecifyShape` `Op`.
```
```suggestion
  - An `Ellipsis` (`...`) in the last position of `shape` or `dims` can be used as short-hand notation for implied dimensions. `size=(3,)` gives the same result as `shape=(3, ...)` but doesn't add a `SpecifyShape` `Op`.
```

OriolAbril commented on pymc3/distributions/distribution.py:
```diff
@@ -46,12 +43,18 @@
 
 PLATFORM = sys.platform
 
+Shape = Union[int, Tuple[Union[str, type(Ellipsis)]], Variable]
```
I understood that `Tuple` from typing is used for fixed length tuples, so that would indicate a length 2 tuple of (str/ellipsis, variable). I am not too sure about anything related to type hints though.

OriolAbril commented on pymc3/distributions/distribution.py:
```diff
@@ -122,13 +125,112 @@ def logcdf(op, var, rvs_to_values, *dist_params, **kwargs):
+    if not isinstance(size, (type(None), int, list, tuple)):
+        raise ValueError("The `size` parameter must be an int, list or tuple.")
+
+    # Warn about backwards-compatible but discouraged parametrization
```
we decided against that so I think this and the comment below about int size should be removed

OriolAbril commented on pymc3/model.py:
```diff
@@ -913,20 +946,138 @@ def shape_from_dims(self, dims):
+            A symbolic scalar of the dimensions length.
+            Defaults to ``aesara.shared(len(values))``.
+        """
+        if name in {"draw", "chain"}:
```
We set that up to avoid name clashes, but it could be added here for completeness, some arviz functions like loo need to flatten the chain and draw dims, and use `"__sample__"` for this dim result of stacking chain and draw together.

OriolAbril commented on pymc3/model.py:
```diff
@@ -913,20 +946,138 @@ def shape_from_dims(self, dims):
+            )
+        if name in self.coords:
+            if not values.equals(self.coords[name]):
+                raise ValueError("Duplicate and incompatiple coordinate: %s." % name)
```
```suggestion
                raise ValueError(f"Duplicate and incompatiple coordinate: {name}.")
```

OriolAbril commented on pymc3/model.py:
```diff
@@ -913,20 +946,138 @@ def shape_from_dims(self, dims):
+            New values for the shared variable.
+        coords : optional, dict
+            New coordinate values for dimensions of the shared variable.
+            Must be provided for all named dimensions that change in length.
```
I think this is actually only for dims that change in length _and had coordinate values already set previously_. Otherwise the gets return none and all ifs related to coords are not entered

OriolAbril reviewed (dismissed):
These improvements on dims are amazing and I personally think most user code will soon use mostly dims instead of either of shape or size.

michaelosthege committed:
Allow parametrization through either shape, dims or size

Always treats `size` as being in addition to dimensions implied by RV parameters.

All resizing beyond parameter-implied dimensionality is done from:
- `shape` or `size` in `Distribution.dist()`
- `dims` or `observed` in `Distribution.__new__`
and only in those two places.

Closes #4552.

michaelosthege committed:
Automatically add SpecifyShape Op when full-length shape is given

brandonwillard commented on RELEASE-NOTES.md:
```diff
@@ -8,11 +8,18 @@
 - The `CAR` distribution has been added to allow for use of conditional autoregressions which often are used in spatial and network models.
+- The dimensionality of model variables can now be parametrized through either of `shape`, `dims` or `size` (see [#4625](https://github.com/pymc-devs/pymc3/pull/4625)):
+  - With `shape` the length of dimensions must be given numerically or as scalar Aesara `Variables`. Using `shape` restricts the model variable to the exact length and re-sizing is no longer possible.
+  - `dims` is arguably the most elegant parametrization, because it allows you to resize `pm.Data` variables and leads to well defined coordinates in `InferenceData` objects.
```
Let's not make qualitative statements about these features, because, for instance, `dims` is not elegant&mdash;or even relevant&mdash;in any sense when there isn't a `pm.Data` object or dimension labels involved.

Language like this is likely to lead people down unnecessary paths, where&mdash;for example&mdash;they believe that `dims` is the "best" approach and that they need to add dimension labels to their models and start using `pm.Data` when they really don't.

OriolAbril replied:
I agree changelog is not the best place for qualitative comments, but I want to note that `dims` can be used without any `pm.Data` with no problem whatsoever. They are related yet completely independent functionality.

`pm.Data` is related to shared data and being able to change observations/predictors without recompiling the model and keeping track of what goes in `constant_data` group of InferenceData. 

`dims` is about specifying `size` or `shape` with names instead of numbers, together with a mapping of name->dim length for everything to work. You can use `coords` in `pm.Model` to define the mapping or use only implied dims the first time a new dimension is used to define that mapping _in addition_ to using `pm.Data` to define the mapping.

michaelosthege replied:
I was under the impression that the word "arguably" phrases it more like a recommendation, but I'm also not a native speaker.
Anyway this part of the sentence was removed in the latest diff.

brandonwillard commented on pymc3/distributions/distribution.py:
```diff
@@ -141,40 +242,124 @@ def __new__(cls, name, *args, **kwargs):
+        n_implied = rv_out.ndim
+
+        # `dims` are only available with this API, because `.dist()` can be used
+        # without a modelcontext and dims are not yet tracked at the Aesara level.
```
Nothing like `dims` will be tracked at the Aesara level; remember, Aesara is basically a wrapper for NumPy and SciPy and does not cover what is essentially Pandas functionality.

That part of the comment should be removed.

OriolAbril replied:
Then we should close https://github.com/pymc-devs/aesara/issues/352 which is what I think motivated this comment.

michaelosthege replied:
Yes, that's related. I removed the "yet" so it no longer implies anything about the future of dim handling. (Nevertheless I think that tracking dims at the tensor level is something worth to consider.)

brandonwillard commented on pymc3/distributions/distribution.py:
```diff
@@ -141,40 +242,125 @@ def __new__(cls, name, *args, **kwargs):
+        if dims is not None:
+            if Ellipsis in dims:
+                # Auto-complete the dims tuple to the full length
+                dims = (*dims[:-1], *[None] * rv_out.ndim)
```
Is the `RandomVariable` created first, and then resized, only because `rv_out.ndim` might be needed in this step?

Regardless, `rv_out.ndim` is obtainable without first creating the underlying `RandomVariable`.

brandonwillard commented on pymc3/distributions/distribution.py:
```diff
@@ -141,40 +242,125 @@ def __new__(cls, name, *args, **kwargs):
+            # were not yet known to the model, such that they can be used for size further downstream.
+            for di, dname in enumerate(dims[n_resize:]):
+                if not dname in model.dim_lengths:
+                    model.add_coord(dname, values=None, length=rv_out.shape[n_resize + di])
```
Again, this logic needs to be factored out into a simple and clear function that&mdash;for instance&mdash;computes an appropriate value of `size` (to be passed to `Distribution.dist`) using values from `dims` or `shape`.

This is especially important given that this code also updates the `model`'s coordinates, making it easily more relevant as a method in `Model`.

In other words, it's best to keep the `Model`-specific logic **in `Model`**, and not spread out across `Distribution` _and_ `Model`.

Also, the steps related to `observed` are apparently duplicating some of the steps taken later in `Model.make_obs_var`.  This might be yet another good reason for moving this logic into `Model`.

brandonwillard commented on pymc3/distributions/distribution.py:
```diff
@@ -141,40 +242,125 @@ def __new__(cls, name, *args, **kwargs):
+
+        shape, _, size = _validate_shape_dims_size(shape=shape, size=size)
+
+        # Create the RV without specifying size or testval.
+        # The size will be expanded later (if necessary) and only then the testval fits.
+        rv_native = cls.rv_op(*dist_params, size=None, **kwargs)
```
[This](https://github.com/pymc-devs/pymc3/blob/45cb4ebf36500e502481bdced6980dd9e630acca/pymc3/aesaraf.py#L156) demonstrates a way to obtain the symbolic shape for a `RandomVariable` without actually creating one.

brandonwillard commented on pymc3/model.py:
```diff
@@ -826,6 +837,27 @@ def basic_RVs(self):
+        """
+        return self._RV_dims
+
+    @property
+    def coords(self) -> Dict[str, Union[Sequence, None]]:
+        """Coordinate values for model dimensions."""
+        return self._coords
+
+    @property
+    def dim_lengths(self) -> Dict[str, Tuple[Variable, ...]]:
+        """The symbolic lengths of dimensions in the model.
+
+        The values are typically instances of ``TensorVariable`` or ``ScalarSharedVariable``.
+        """
+        return self._dim_lengths
```
Why not add the symbolic dimension variables to the existing `Model._coords`?  In other words, the current approach uses two maps, but one is sufficient.

michaelosthege replied:
Those are different maps. One is not sufficient: `Model._coords` are the **vectors** with coordinate values. `Model.dim_lengths` are scalars. They are not the same and `Model.coords` was there before, so it should not be changed.

brandonwillard commented on pymc3/model.py:
```diff
@@ -913,20 +945,144 @@ def shape_from_dims(self, dims):
+        shared_object = self[name]
+        if not isinstance(shared_object, SharedVariable):
+            raise TypeError(
+                f"The variable `{name}` must be defined as `pymc3.Data` inside the model to allow updating. "
```
`SharedVariable`s and `pymc3.Data` are not the same things.  If all this logic works with the more general `SharedVariable`s, then it needs to be worded that way.  If not, this `isinstance` might not be specific enough.

brandonwillard commented on pymc3/tests/test_data_container.py:
```diff
@@ -272,9 +296,49 @@ def test_explicit_coords(self):
+    def test_symbolic_coords(self):
+        """
+        In v4 dimensions can be created without passing coordinate values.
+        Their length is then automatically linked to the corresponding Tensor dimension.
```
```suggestion
        Their lengths are then automatically linked to the corresponding Tensor dimension.
```

brandonwillard commented on pymc3/tests/test_data_container.py:
```diff
@@ -272,9 +296,49 @@ def test_explicit_coords(self):
+            intensity.set_value(floatX(np.ones((4, 5))))
+            assert pmodel.dim_lengths["row"].eval() == 4
+            assert pmodel.dim_lengths["column"].eval() == 5
+        pass
```
```suggestion
```

brandonwillard commented on pymc3/tests/test_data_container.py:
```diff
@@ -272,9 +296,49 @@ def test_explicit_coords(self):
+                ShapeError, match="was initialized from 'n' which is not a shared variable"
+            ):
+                pmodel.set_data("inhabitants", [1, 2, 3, 4])
+        pass
```
```suggestion
```

brandonwillard commented on pymc3/tests/test_logp.py:
```diff
@@ -70,23 +70,23 @@ def test_logpt_basic():
     rng = aesara.shared(np.random.RandomState(232), borrow=True)
 
-    a = Normal.dist(mu, sigma, size=size, rng=rng)
+    a = Normal.dist(mu, sigma, shape=shape, rng=rng)
```
Again, `shape` should have its own tests.  This test has&mdash;and should have&mdash;nothing to do with the `shape` feature.

These changes need to be removed.

michaelosthege replied:
Then why should it have anything to do with the `size` feature? It already parametrized the RV to have shape `size` through its `mu` parameter only to replicate it again by the same tuple.
In the 4th parametrization this creates an RV with full shape `(5, 5, 5, 5)` which is unnecessary for what's being tested. Also counterintuitively the other parametrizations don't test with a scalar RV, but always with a `(5, 5)`-shaped one.
If anything the `size` kwarg could be removed entirely.

brandonwillard commented on pymc3/tests/test_model.py:
```diff
@@ -461,33 +461,29 @@ def test_make_obs_var():
     sparse_output = fake_model.make_obs_var(fake_distribution, sparse_input, None, None)
+    assert sparse_output == fake_distribution
+    assert sparse.basic._is_sparse_variable(sparse_output.tag.observations)
     del fake_model.named_vars[fake_distribution.name]
+
+    # Here the RandomVariable is split into observed/imputed and a Deterministic is returned
     masked_output = fake_model.make_obs_var(fake_distribution, masked_array_input, None, None)
+    assert masked_output != fake_distribution
     assert not isinstance(masked_output, RandomVariable)
-
-    # Ensure that the missing values are appropriately set to None
-    for func_output in [dense_output, sparse_output]:
-        assert isinstance(func_output.owner.op, RandomVariable)
-
-    # Ensure that the Aesara variable names are correctly set.
-    # Note that the output for masked inputs do not have their names set
-    # to the passed value.
-    for func_output in [dense_output, sparse_output]:
-        assert func_output.name == input_name
-
-    # Ensure the that returned functions are all of the correct type
-    assert isinstance(dense_output.tag.observations, TensorConstant)
-    assert sparse.basic._is_sparse_variable(sparse_output.tag.observations)
-
-    # Masked output is something weird. Just ensure it has missing values
+    # Ensure it has missing values
```
Why was this test changed?  More importantly, what does it have to do with the method being tested (i.e. `Model.make_obs_var`)?

michaelosthege replied:
The test relied on the recreation of the `fake_distribution` by resizing in `make_obs_var`.
Without that recreation, the dense observations attached to `fake_distribution` changed to sparse observations before before the asserts.

Note that I added `assert dense_output == fake_distribution` which would have failed before, when `make_obs_var` resized the RV if it did not match the observations shape.

brandonwillard commented on pymc3/tests/test_sampling.py:
```diff
@@ -385,11 +385,10 @@ def test_shared_named(self):
             )
             theta = pm.Normal(
-                "theta", mu=at.dot(G_var, theta0), tau=np.atleast_2d(1e20), size=(1, 1)
+                "theta", mu=at.dot(G_var, theta0), tau=np.atleast_2d(1e20), shape=(1, 1)
```
```suggestion
                "theta", mu=at.dot(G_var, theta0), tau=np.atleast_2d(1e20), size=(1, 1)
```

michaelosthege replied:
Again, this test has nothing to do with `size`. All other variables here are just 2-dimensional and there is no need to make it 4-dimensional.

brandonwillard commented on pymc3/tests/test_sampling.py:
```diff
@@ -401,11 +400,10 @@ def test_shared_unnamed(self):
             )
             theta = pm.Normal(
-                "theta", mu=at.dot(G_var, theta0), tau=np.atleast_2d(1e20), size=(1, 1)
+                "theta", mu=at.dot(G_var, theta0), tau=np.atleast_2d(1e20), shape=(1, 1)
```
```suggestion
                "theta", mu=at.dot(G_var, theta0), tau=np.atleast_2d(1e20), size=(1, 1)
```

michaelosthege replied:
Same as above.

brandonwillard commented on pymc3/tests/test_sampling.py:
```diff
@@ -417,11 +415,10 @@ def test_constant_named(self):
             )
             theta = pm.Normal(
-                "theta", mu=at.dot(G_var, theta0), tau=np.atleast_2d(1e20), size=(1, 1)
+                "theta", mu=at.dot(G_var, theta0), tau=np.atleast_2d(1e20), shape=(1, 1)
```
```suggestion
                "theta", mu=at.dot(G_var, theta0), tau=np.atleast_2d(1e20), size=(1, 1)
```

michaelosthege replied:
Same as above.

brandonwillard commented on pymc3/tests/test_sampling.py:
```diff
@@ -936,14 +933,14 @@ def test_ignores_observed(self):
             with pm.Model():
-                mu = pm.Gamma("mu", 3, 1, size=1)
-                goals = pm.Poisson("goals", mu, size=shape)
+                mu = pm.Gamma("mu", 3, 1)
```
You've removed `size=1`, but are you sure that this wasn't supposed to be part of the test?

Also, why remove it in the first place?

michaelosthege replied:
Making `mu` a length-1 vector and also passing `size=shape` in `Poisson` makes the resulting RV `(*shape, 1)`.
Consequently the assert of trace shape fails.

Again, this test only worked before, because the effect of `size` were inconsistent.

michaelosthege replied:
After our misunderstanding I'm now revisiting this test to evaluate if it should be reverted.
So what the test did before my changes:
1. Create a `(1,)` `Gamma` RV
2. Use that as the `mu` parameter in a `Poisson`, but then set the `size` to `(2,)`, `10, 2)` and so on,

So essentially it did not respect that `mu` was actually a length 1 _vector_ and treated it as if it was a scalar.
Was it really the intention of the test to verify that a lengh-1 vector input parameter is treated like a scalar, or is that just a leftover from our old shape handling before #4214, when `size=1` became scalar?

brandonwillard commented on pymc3/tests/test_sampling.py:
```diff
@@ -936,14 +933,14 @@ def test_ignores_observed(self):
-                mu = pm.Gamma("mu", 3, 1, size=1)
-                goals = pm.Poisson("goals", mu, size=shape)
+                mu = pm.Gamma("mu", 3, 1)
+                assert mu.eval().shape == ()
```
Why are you sampling this variable and then checking its shape?  What does that have to do with this test, which is apparently about the shapes of the subsequently produced traces?

brandonwillard commented on pymc3/tests/test_sampling.py:
```diff
@@ -971,14 +968,14 @@ def test_multivariate2(self):
-            a = pm.Uniform("a", lower=0, upper=1, size=10)
-            b = pm.Binomial("b", n=1, p=a, size=10)
+            a = pm.Uniform("a", lower=0, upper=1, size=5)
+            b = pm.Binomial("b", n=1, p=a, size=7)
 
         model.default_rng.get_value(borrow=True).seed(232093)
 
         b_sampler = aesara.function([], b)
         avg = np.stack([b_sampler() for i in range(10000)]).mean(0)
-        npt.assert_array_almost_equal(avg, 0.5 * np.ones((10,)), decimal=2)
+        npt.assert_array_almost_equal(avg, 0.5 * np.ones((7, 5)), decimal=2)
```
Again, why were these changed?

michaelosthege replied:
This is another example where the previous implementation did not actually treat `size` as being _in addition to_ implied dimensions:
```
a = Uniform(..., size=5)         # this is now a vector
b = Binomial(..., p=a, size=7)   # ndim=1 is already implied through `p`; size scales it up to a matrix!
```
Using non-equal sizes like the prime numbers 5 and 7 makes the test easier to debug in case some dimensions are flipped.

brandonwillard commented on pymc3/tests/test_shape_handling.py:
```diff
@@ -28,6 +31,8 @@
     to_tuple,
 )
 
+_log = logging.getLogger(__file__)
```
Why would we need logging in a test module?  If this is for local debugging, then it should be removed; otherwise, there's no reason to add noise to the output of our tests.

michaelosthege replied:
I used it just for logging and forgot to remove it before committing. Will do.

brandonwillard commented on pymc3/distributions/distribution.py:
```diff
@@ -46,12 +43,18 @@
     "NoDistribution",
 ]
 
+_log = logging.getLogger(__file__)
```
Is this used for anything?

brandonwillard commented on pymc3/tests/test_shape_handling.py:
```diff
@@ -219,3 +224,169 @@ def test_sample_generate_values(fixture_model, fixture_sizes):
+                        assert rv.eval().shape == batch_shape + param_shape
+                    else:
+                        raise NotImplementedError("Invalid test case parametrization.")
+        pass
```
```suggestion
```

brandonwillard commented on pymc3/tests/test_shape_handling.py:
```diff
@@ -219,3 +224,169 @@ def test_sample_generate_values(fixture_model, fixture_sizes):
+            agedata.set_value([1, 2, 3, 4])
+            # The change should propagate all the way through
+            assert effect.eval().shape == (4,)
+        pass
```
```suggestion
```

brandonwillard commented on pymc3/tests/test_shape_handling.py:
```diff
@@ -219,3 +224,169 @@ def test_sample_generate_values(fixture_model, fixture_sizes):
+            y = pm.Normal("y", mu=x, size=2, dims=("third", "first", "second"))
+            assert "third" in pmodel.dim_lengths
+            assert y.eval().shape() == (2, 1, 4)
+        pass
```
```suggestion
```

brandonwillard commented on pymc3/tests/test_shape_handling.py:
```diff
@@ -219,3 +224,169 @@ def test_sample_generate_values(fixture_model, fixture_sizes):
+            assert x.eval().shape == (3, 2)
+            assert y.eval().shape == (3, 2)
+            assert z.eval().shape == (3, 2)
+        pass
```
```suggestion
```

brandonwillard commented on pymc3/tests/test_shape_handling.py:
```diff
@@ -219,3 +224,169 @@ def test_sample_generate_values(fixture_model, fixture_sizes):
+        assert pm.Normal.dist(mu=mu, shape=(5, 3)).eval().shape == (5, 3)
+        assert pm.Normal.dist(mu=mu, shape=(7, ...)).eval().shape == (7, 3)
+        assert pm.Normal.dist(mu=mu, size=(4,)).eval().shape == (4, 3)
+        pass
```
```suggestion
```

brandonwillard commented on pymc3/tests/test_shape_handling.py:
```diff
@@ -219,3 +224,169 @@ def test_sample_generate_values(fixture_model, fixture_sizes):
+        assert pm.Uniform.dist(2, [4, 5], shape=[3, 2]).eval().shape == (3, 2)
+        with pm.Model(coords=dict(town=["Greifswald", "Madrid"])):
+            assert pm.Normal("n2", mu=[1, 2], dims=("town",)).eval().shape == (2,)
+        pass
```
```suggestion
```

brandonwillard commented on pymc3/tests/test_step.py:
```diff
@@ -982,7 +982,7 @@ def test_linalg(self, caplog):
             a = at.switch(a > 0, np.inf, a)
             b = at.slinalg.solve(floatX(np.eye(2)), a)
-            Normal("c", mu=b, size=2, testval=floatX(np.r_[0.0, 0.0]))
+            Normal("c", mu=b, shape=(2,), testval=floatX(np.r_[0.0, 0.0]))
```
```suggestion
            Normal("c", mu=b, size=2, testval=floatX(np.r_[0.0, 0.0]))
```

michaelosthege replied:
Same as with other examples: The way `size` was used before the `testval` shape does not match. Like with other examples, this was likely an unintended change of behavior from the original test when it was switched to `size`.

brandonwillard commented on pymc3/tests/test_transforms.py:
```diff
@@ -274,11 +274,11 @@ def test_chain_jacob_det():
 class TestElementWiseLogp(SeededTest):
-    def build_model(self, distfam, params, size, transform, testval=None):
+    def build_model(self, distfam, params, *, size=None, shape=None, transform=None, testval=None):
         if testval is not None:
             testval = pm.floatX(testval)
         with pm.Model() as m:
-            distfam("x", size=size, transform=transform, testval=testval, **params)
+            distfam("x", size=size, shape=shape, transform=transform, testval=testval, **params)
```
All of the changes in this module need to be reverted; they have nothing to do with the `shape` or `dims` feature, nor should they.

michaelosthege replied:
All of the changes in this module are because many of these tests are actually parametrized by `shape`.
Consider this test case from before the change:
```python
    @pytest.mark.parametrize(
        "a,b,size",
        [
            (1.0, 1.0, (2,)),                               # scalar is implied; 2 replicate dimensions
            (np.ones(3), np.ones(3), (4, 3)),               # (3,) is implied; (4, 3) replicate dimensions
        ],
    )
    def test_beta_ordered(self, a, b, size):
        testval = np.sort(np.abs(np.random.rand(*size)))    # but the testval is at most 2-dimensional
        model = self.build_model(
            pm.Beta,
            {"alpha": a, "beta": b},
            size=size,                     # in the second parametrize case the RV is expanded to (4, 3, 3)
            testval=testval,               # but the testval is only (4, 3)
            transform=tr.Chain([tr.logodds, tr.ordered]),
        )
        self.check_vectortransform_elementwise_logp(model, vect_opt=0)
```

Comparing with `master` we can see that the test was up to 2-dimensional to begin with. The change to `size` had changed the behaviour of the test - I'm just reverting it to what was originally intended & tested.
https://github.com/pymc-devs/pymc3/blob/2dee9dafb3541043ae23b66fe9262e70fbd2769a/pymc3/tests/test_transforms.py#L405-L421

Yes, the test could be refactored to use `size`, by either replicating the `shape`←→`size` logic, or parametrizing the test with both `shape` and `size` manually. That would be a lot of unnecessary work and make the tests harder to work with!
Since this PR is about restoring `shape` as a backwards-compatible and fully supported way to parametrize RVs I do not see the point in making our lives harder by notoriously avoiding it in tests like this one.

brandonwillard replied:
I checked the `testval`s for this test in `v3` and `v4` and they're both the same.  The symbolic types for the resulting variables are also the same between `v3` and `v4`, so I don't understand what's changed here&mdash;aside from the switch to `size`.

More importantly, this is a scalar variable, so `size` and `shape` should be the same.  If they're not, then that's an altogether different issue that needs to be addressed separately.

Regardless, the change here is still an unnecessary reversion that forces these tests to use an irrelevant and newly added feature.

brandonwillard reviewed (changes requested):
There are a lot of unnecessary `shape` keyword changes in unrelated tests that need to be reverted.  As well, there seem to be some similarly unrelated non-`shape` test changes that need to be reverted.

Otherwise, the logic in `Distribution.[__new__|dist]` is still more complicated than it needs to be and should be better organized and distributed between the `Model` and `Distribution` classes.  

Under these changes, `Distribution` is taking on steps that are&mdash;and were&mdash;performed in `Model`, and this is likely the cause of the current failures.  Without an unusually compelling reason for this previously `Model`-specific logic to be relocated to `Distribution`, these changes are not sound.

For instance, the steps in `Distribution` that are performed based on the value of `observed` are actually `Model`-exclusive, because observations are a `Model`-only concept and feature.  By moving&mdash;or adding&mdash;parts of that logic into `Distribution`, we fragment our logic and violate the abstractions that underlie these classes.  The end result is code that's harder to follow&mdash;because the underlying abstractions are much less relevant to the actual implementation&mdash;and exceedingly more difficult to maintain and extend, due to a less manageable back-and-forth between the two classes.


In the end, these changes should only require a single call in `Distribution.__new__` to a function that&mdash;at maximum&mdash;takes the `size`, `shape`, `dims`, `rv_op`, and distribution parameters as arguments, determines the appropriate `size` to be passed to `Distribution.dist`, and that's all.

This function could be a method provided by `Model`, which would make it reasonable to also perform any relevant `Model` updates that involve `dims` within it.

The only step that seems like it could/should be in `Distribution` is the application of `specify_shape` when `shape` is used, but I'm not seeing that in this diff.

michaelosthege commented:
The test about #4652 `XPASS`ed: https://github.com/pymc-devs/pymc3/pull/4625/checks?check_run_id=2394141812#step:7:3572

It shouldn't have done that - looks like https://github.com/pymc-devs/pymc3/blob/v4/pyproject.toml#L2 isn't working.

With respect to the other changes...

Reasons for why some tests were changed are thoroughly commented above (see threads).

Regarding other options to obtain implied dimensionality information, I'd like to quote @brandonwillard:
> it's generally cheap to make these graphs, so you can always construct one just for information
> also, graph construction isn't a "run-time" thing, so it doesn't really affect performance

So that's how I implemented it. There might be other ways, but given that it's cheap, that it works and that it restores the important backwards compatibility while bringing more flexibility I don't think we should delay this PR any further.

I think the shape/dims/size/size_from_observed logic from the two `Distribution` APIs should not be split across `distribution.py`/`model.py`. Not only do they reference the same helper functions, but the `Distribution` class is also the point of entry. After moving the determination of `size` from `observed` also into `Distribution.__new__` the `model.py` is always given RVs with the correct dimensionality already and doesn't even import `change_rv_size`.
Factoring out the logic within `Distribution.__new__` can possibly be addressed in a follow-up PR that makes `dims` work _alongside_ `shape`/`size` (see PR description).

So from my point of view this PR is all good cc @brandonwillard @OriolAbril @twiecki

OriolAbril reviewed (approved):
I mostly agree with @michaelosthege last comment, but my opinion on model/distribution topics probably should not matter much as I only have a loose understanding their internals. 

I do want to add an extra point on that however, as a relative sideliner. I feel that without a clear/detailed roadmap (that I know of) and with things like https://github.com/pymc-devs/aesara/issues/352 on the air, I am not sure it's possible at this point to know what is the correct level of abstraction for each thing. There seem to be arguments to move `dims` in either direction, towards `dist` and a better integration with Aesara or towards `Model`.

And last but definitely not least, I want to say that the more I see about the capabilities that `v4` is improving/adding/extending the more amazed I am with it. Thanks to everyone that is pushing hard on this and leading the effort.

twiecki commented:
I agree, I think this looks great. Agree with @OriolAbril that at this point it's not clear what the right level of abstraction is and I don't see a clear winner. We can always revisit when we find out later that we can improve things.
